### PR TITLE
Allow custom ExternArgument in extern declaration

### PR DESCRIPTION
**Pull request by anuragm**
This is useful if people want to specify ArrayReferenceType as an argument to an extern function. ArrayReferenceType require an explicit Access modifier (mutable or readonly). Since this is just one edge case, we can let users still specify just the type for all other extern declarations.

I have added a test that demos its usage.

### CODE REVIEW

File: oqpy/subroutines.py [hunk 1]
```diff
@@ -194,7 +194,7 @@ def declare_extern(
     arg_types = list(zip(*(args)))[1] if args else []
     extern_decl = ast.ExternDeclaration(
         ast.Identifier(name),
-        [ast.ExternArgument(type=t) for t in arg_types],
+        [ast.ExternArgument(type=t) if isinstance(t, ast.ClassicalType) else t for t in arg_types],
```
[the remark this answers is not in the document]
Comment (anuragm) [reply oqpy/subroutines.py]: Ok. I went ahead and added a data class called `OqpyArgument` that stores all the information required for an argument. I also added an `arrayreference_` function which serves a similar function to things like `int_` which allow us to quickly add extern declaration.

File: oqpy/subroutines.py [hunk 2]
```diff
@@ -30,13 +31,26 @@
...
-__all__ = ["subroutine", "declare_extern", "declare_waveform_generator"]
+__all__ = ["subroutine", "declare_extern", "declare_waveform_generator", "OqpyArgument"]
 
 SubroutineParams = [oqpy.Program, VarArg(AstConvertible)]
 
 FnType = TypeVar("FnType", bound=Callable[..., Any])
 
 
+@dataclass
+class OqpyArgument:
```
[the remark this answers is not in the document]
Comment (anuragm) [reply oqpy/subroutines.py]: Done.

File: oqpy/subroutines.py [hunk 3]
```diff
@@ -30,13 +31,26 @@
...
 
-__all__ = ["subroutine", "declare_extern", "declare_waveform_generator"]
+__all__ = ["subroutine", "declare_extern", "declare_waveform_generator", "OqpyArgument"]
 
 SubroutineParams = [oqpy.Program, VarArg(AstConvertible)]
 
 FnType = TypeVar("FnType", bound=Callable[..., Any])
 
 
+@dataclass
```
[the remark this answers is not in the document]
Comment (anuragm) [reply oqpy/subroutines.py:41]: Yes, using a dataclass let's us convert from literal string arguments to enum argument inside `unpack()`. Not sure how to do it with `NamedTuple` without exposing `ast.AccessControl` directly, or moving the functionality of `unpack()` where-ever the `NamedTuple` is consumed.

File: oqpy/subroutines.py [hunk 4]
```diff
@@ -30,13 +31,26 @@
...
+class OQPyArgument:
+    """An oqpy argument to extern declaration.."""
+
+    name: str
+    dtype: ast.ClassicalType
+    access: Literal["readonly", "mutable"] | None = None
```
Comment (braised-babbage) [inline oqpy/subroutines.py:47]: This follows the precedent established by the OpenQASM reference AST, so it's fine to have here. But I will say that I think the underlying premise is slightly mistaken. Access modifiers or type qualifiers make most sense as part of the type itself. In a subroutine, if you ask "what is the type of this variable" it's nice to be able to say "a mutable array reference".
